### PR TITLE
feat: [PL-39460]: api def for secret decryption

### PR DESCRIPTION
**Pull request by xingchi-jin**
## Describe your changes
This PR contains only proto definitions
1. API definition for GRPC api consumed by module teams
2. API definition for task acquire regarding secret decryption. Spec first ! 

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52974)

### CODE REVIEW

File: 955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto [hunk 1]
```diff
@@ -29,4 +29,225 @@ message SecretConfig {
...
+
+message BijouSecret {
+  EncryptedDataRecord encrypted_data = 1;
+  EncryptionConfig config = 2;
+}
+
+/*
+  Data model of encrypted records acquired by delegate for decryption
+*/
+message EncryptedDataRecord {
```
Comment (gn-harness) [inline 955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto:42]: How did we arrive at these 13 fields? In the java object I see we also have fields like parameters, parents, enabled, fileSize etc.

Can you please an approval from Nishant/Meenakshi too?

**Review by gn-harness — APPROVED** [review]

**Review by mdmilic — COMMENTED** [review]
> Reviewed 1 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52974)* status: 4 file(s) to review. 1 of 2 approvals obtained. 

___
*[`920-delegate-service-beans/src/main/proto/io/harness/delegate/secrets.proto` line 12 at r1](https://reviewable.io/reviews/harness/harness-core/52974#-NebhFKPDH2bUwfMb68h:-NebhFKPDH2bUwfMb68i:bidjhwm) ([raw file](https://github.com/harness/harness-core/blob/3c435bbf812eafc861b75ba1825f49560f6ff068/920-delegate-service-beans/src/main/proto/io/harness/delegate/secrets.proto#L12)):*
> ```proto
> option go_package = "920_delegate_service_beans";
> 
> message Secrets {
> ```

Any particular reason for this wrapper? I think it would just make code more verbose, do you expect something else to be part of `Secrets` except for list of `Secret`
___
*[`955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto` line 42 at r1](https://reviewable.io/reviews/harness/harness-core/52974#-Nebi7lh0X9x_pM7rtQ9:-Nebi7lh0X9x_pM7rtQA:bqaqudn) ([raw file](https://github.com/harness/harness-core/blob/3c435bbf812eafc861b75ba1825f49560f6ff068/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto#L42)):*
> ```proto
>   Data model of encrypted records acquired by delegate for decryption
> */
> message EncryptedDataRecord {
> ```

Whoa spidey sense tingling. This doesn't seem like a good approach. What alternatives do we have? 
Obviously we don't want to serialize to Kryo, can we maybe push more left (towards the secret service), or we think that's not a good idea because delegate needs to understand these? How can we make it more pluggable to be able to add more secret providers? How about 3rd party secret providers? Why does DMS need to know anything about these secrets (why does Secret Manager need to serialize them to real data, we deserialize in DMS and then serialize to Delegate data etc...)?
If we still end up doing this, can we scrutinize these messages because they have years of technical debt that we don't need to import into our brand new APIs.

**Review by xingchi-jin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52974)* status: Some of the checks are too old. 4 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto` line 42 at r1](https://reviewable.io/reviews/harness/harness-core/52974#-Nebi7lh0X9x_pM7rtQ9:-NeiBLIKCGcscx1Q3sQO:b-wy6xg6) ([raw file](https://github.com/harness/harness-core/blob/3c435bbf812eafc861b75ba1825f49560f6ff068/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto#L42)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

Whoa spidey sense tingling. This doesn't seem like a good approach. What alternatives do we have? 
Obviously we don't want to serialize to Kryo, can we maybe push more left (towards the secret service), or we think that's not a good idea because delegate needs to understand these? How can we make it more pluggable to be able to add more secret providers? How about 3rd party secret providers? Why does DMS need to know anything about these secrets (why does Secret Manager need to serialize them to real data, we deserialize in DMS and then serialize to Delegate data etc...)?
If we still end up doing this, can we scrutinize these messages because they have years of technical debt that we don't need to import into our brand new APIs.
</blockquote></details>

i understand your concern and where you come from.
Best things takes time to happen. What you're describing is to improve the secret manager framework's design overall. Make it pluggable and make it easy to add new SMs in the future.
Here the problem to solve, is to make task pluggable first. For this goal, making secret decryption API spec-first is the target. 

After this is achieved, we think about making the SM framework pluggable of adding new SMs. Till then, only the SMs that has been implemented in code will be supported, just like today. 

 Please suggest your alternative, or AIs. It can help us overall improve.

**Review by xingchi-jin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52974)* status: Some of the checks are too old. 4 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto` line 42 at r1](https://reviewable.io/reviews/harness/harness-core/52974#gh-1327916996:-NeiCV_b4vUNgM1UZ7nM:b-12n1yc) ([raw file](https://github.com/harness/harness-core/blob/3c435bbf812eafc861b75ba1825f49560f6ff068/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto#L42)):*
<details><summary><i>Previously, gn-harness (Gaurav Nanda) wrote…</i></summary><blockquote>

How did we arrive at these 13 fields? In the java object I see we also have fields like parameters, parents, enabled, fileSize etc.

Can you please an approval from Nishant/Meenakshi too?
</blockquote></details>

It comes from the EncryptedRecordData java entity

**Review by xingchi-jin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52974)* status: Some of the checks are too old. 4 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto` line 42 at r1](https://reviewable.io/reviews/harness/harness-core/52974#-Nebi7lh0X9x_pM7rtQ9:-Nejh5le6LEtBaETxXH9:blvvcpg) ([raw file](https://github.com/harness/harness-core/blob/3c435bbf812eafc861b75ba1825f49560f6ff068/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto#L42)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

i understand your concern and where you come from.
Best things takes time to happen. What you're describing is to improve the secret manager framework's design overall. Make it pluggable and make it easy to add new SMs in the future.
Here the problem to solve, is to make task pluggable first. For this goal, making secret decryption API spec-first is the target. 

After this is achieved, we think about making the SM framework pluggable of adding new SMs. Till then, only the SMs that has been implemented in code will be supported, just like today. 

 Please suggest your alternative, or AIs. It can help us overall improve.
</blockquote></details>

Marko and I discussed this morning, and now i understand better, so ignore my previous comment
The comment can be resolved by
1. the API section for secrets to be decrypted should be defined as binary, just like task data. Because secrets decryption is like a layer on top of delegate task routing layer. Which i completely agree. 
2. In k8s init task, the secrets belongs to different containers. We should have secret id defined in the container spec 

I will do the above two changes

**Review by nishantsaini20 — COMMENTED** [review]
> Reviewed 2 of 4 files at r1, 2 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52974)* status: 4 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto` line 42 at r1](https://reviewable.io/reviews/harness/harness-core/52974#gh-1327916996:-NepzPPF68UbfLCtxvf5:bljiv30) ([raw file](https://github.com/harness/harness-core/blob/3c435bbf812eafc861b75ba1825f49560f6ff068/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto#L42)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

It comes from the EncryptedRecordData java entity
</blockquote></details>

private Set<EncryptedDataParams> parameters; from EncryptedRecordData is missing here.
___
*[`955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto` line 38 at r2](https://reviewable.io/reviews/harness/harness-core/52974#-Nepzd9i4wriqO8ScXN8:-Nepzd9i4wriqO8ScXN9:bxkv0qx) ([raw file](https://github.com/harness/harness-core/blob/e321ec53ca054a806e115dc162099a8649dfefbe/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto#L38)):*
> ```proto
>   string encryption_key = 4;
>   bytes encryptedValue = 5;
>   string kmsId = 6;
> ```

nit: Why there is mix of camel case and snake case in the field names?

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52974)* status: 4 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained.

**Comment by piyush-harness** [comment]
*[`955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto` line 38 at r2](https://reviewable.io/reviews/harness/harness-core/52974#-Nepzd9i4wriqO8ScXN8:-Netq1ku5TdOG9JVaZit:b32v57) ([raw file](https://github.com/harness/harness-core/blob/e321ec53ca054a806e115dc162099a8649dfefbe/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto#L38)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

nit: Why there is mix of camel case and snake case in the field names?
</blockquote></details>

+1

**Review by piyush-harness — DISMISSED** [review]
> Reviewed 2 of 4 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52974)* status: 1 file(s) to review. 1 change requests. 3. unresolved discussions. 1 of 2 approvals obtained.

**Review by xingchi-jin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52974)* status: 1 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`920-delegate-service-beans/src/main/proto/io/harness/delegate/delegate_task_apis.proto` line 23 at r1](https://reviewable.io/reviews/harness/harness-core/52974#gh-1327654690:-Nf9QwrY5nip37Ee5QmL:b-g87z29) ([raw file](https://github.com/harness/harness-core/blob/3c435bbf812eafc861b75ba1825f49560f6ff068/920-delegate-service-beans/src/main/proto/io/harness/delegate/delegate_task_apis.proto#L23)):*
<details><summary><i>Previously, gn-harness (Gaurav Nanda) wrote…</i></summary><blockquote>

nit: Generally we should not change numbers on the proto fields, it is ok for now as no one is using it
</blockquote></details>

i think it's ok, because it's not used in production or any system. only used in developer space at this moment 
___
*[`955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto` line 42 at r1](https://reviewable.io/reviews/harness/harness-core/52974#gh-1327916996:-Nf9Qq8X2qdHTkVH2jWv:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/3c435bbf812eafc861b75ba1825f49560f6ff068/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto#L42)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

private Set<EncryptedDataParams> parameters; from EncryptedRecordData is missing here.
</blockquote></details>

Done.

**Review by nishantsaini20 — APPROVED** [review]
> Reviewed 1 of 2 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52974)* status: 2 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Review by mdmilic — COMMENTED** [review]
> Reviewed 1 of 4 files at r1, 1 of 3 files at r2, 1 of 2 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52974)* status: 2. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/infra.proto` line 13 at r5](https://reviewable.io/reviews/harness/harness-core/52974#-NfAXg6U7c3kAR6hM8fP:-NfAXg6U7c3kAR6hM8fQ:bm8dgxk) ([raw file](https://github.com/harness/harness-core/blob/813b18f042999bf62d785d98f13f28bdf6dc6fbc/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/infra.proto#L13)):*
> ```proto
> import "955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/plugin.proto";
> import "955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/resource.proto";
> import "955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto";
> ```

I think you probably don't need this import any more and even could delete the entire proto most likely
___
*[`955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto` line 42 at r1](https://reviewable.io/reviews/harness/harness-core/52974#-Nebi7lh0X9x_pM7rtQ9:-NfAY8yx0WO5eJl5bou3:b-s1dwul) ([raw file](https://github.com/harness/harness-core/blob/3c435bbf812eafc861b75ba1825f49560f6ff068/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto#L42)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

Marko and I discussed this morning, and now i understand better, so ignore my previous comment
The comment can be resolved by
1. the API section for secrets to be decrypted should be defined as binary, just like task data. Because secrets decryption is like a layer on top of delegate task routing layer. Which i completely agree. 
2. In k8s init task, the secrets belongs to different containers. We should have secret id defined in the container spec 

I will do the above two changes
</blockquote></details>

You shouldn't need these data models at this place any more. They should be in a separate proto file owned by secret service (not delegate I think - i.e. it's the API spec secret service defines and owns and should DMS should not have to import them).

**Comment by xingchi-jin** [comment]
trigger uts

**Review by xingchi-jin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52974)* status: 5 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto` line 42 at r1](https://reviewable.io/reviews/harness/harness-core/52974#-Nebi7lh0X9x_pM7rtQ9:-NfERlnHBr_LWskKzHuZ:b-o9waz5) ([raw file](https://github.com/harness/harness-core/blob/3c435bbf812eafc861b75ba1825f49560f6ff068/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/secret.proto#L42)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

You shouldn't need these data models at this place any more. They should be in a separate proto file owned by secret service (not delegate I think - i.e. it's the API spec secret service defines and owns and should DMS should not have to import them).
</blockquote></details>

put in another file secret_details.proto

**Review by xingchi-jin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52974)* status: 5 file(s) to review. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/infra.proto` line 13 at r5](https://reviewable.io/reviews/harness/harness-core/52974#-NfAXg6U7c3kAR6hM8fP:-NfES5Pb78G2kfR-th0w:b-hfpv0l) ([raw file](https://github.com/harness/harness-core/blob/b3cdcef96c10782f77a638859931ae9b47da2e3b/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/infra.proto#L13)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

I think you probably don't need this import any more and even could delete the entire proto most likely
</blockquote></details>

removed

**Comment by xingchi-jin** [comment]
trigger uts

**Comment by xingchi-jin** [comment]
trigger uts

**Comment by xingchi-jin** [comment]
trigger ut3

**Comment by xingchi-jin** [comment]
trigger ut9

**Review by nishantsaini20 — APPROVED** [review]
> Reviewed 3 of 5 files at r6, 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52974)* status: 6 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Comment by xingchi-jin** [comment]
trigger ut3

**Pull request merged**